server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -84,6 +85,23 @@ public long readLong() throws IOException {
         }
     }
 
+    @Override
+    public String readString() throws IOException {
+        final int chars = readArraySize();
+        if (slice.hasArray()) {
+            // attempt reading bytes directly into a string to minimize copying
+            final byte[] bytes = slice.array();
+            final int start = slice.position() + slice.arrayOffset();
+            final int limit = slice.limit() + slice.arrayOffset();
+            final int length = calculateByteLengthOfChars(bytes, chars, start, limit);
+            if (length >= 0) {
+                slice.position(length + start - slice.arrayOffset());
+                return new String(bytes, start, length, StandardCharsets.UTF_8);
```

Comment:
If `length == chars` (surely the most common case) I think it's safe to use `ISO_8859_1` instead of `UTF_8`, which skips a few extra checks when constructing the `String`.

Reply:
In fact this is such a common case that I wonder if it'd be even faster just to scan `bytes` looking to see if any of them have the top bit set, falling back to today's more general impl otherwise. Seems like an easier thing for the JIT to optimise into a really tight loop vs all the branching in `calculateByteLengthOfChars`.

Reply:
@original-brownbear WDYT, we've talked about this. Happy to come up with another impl so we can benchmark and compare. In any case, it'll be a lot simpler!

Reply:
Nice idea! I agree with @mosche, if we want to try it this is worth a benchmark :)
I wonder if we could just use jmh here, since what we want to benchmark is pretty well defined (do we use it anywhere in the codebase?)

Reply:
> In fact this is such a common case that I wonder if it'd be even faster just to scan bytes looking to see if any of them have the top bit set, falling back to today's more general impl otherwise. Seems like an easier thing for the JIT to optimise into a really tight loop vs all the branching in calculateByteLengthOfChars.

@DaveCTurner I've optimized the loop as suggest, let me know what you think.